File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.py
# Binary search approach is based on the fact that a rotated sorted array can be divided into two sorted arrays.
# Time complexity: Brute force approach is O(n), Binary search approach is O(log n)
# Space complexity: O(1)
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        start, end = 0, len(nums) - 1
        mid = (start + end) / 2
        while start <= end:
            mid = (start + end) / 2
            if target == nums[mid]:
                return mid
            if nums[start] <= nums[mid]:
                if nums[start] <= target and nums[mid] >= target:
                    end = mid - 1
                else:
                    start = mid + 1
            else:
                if nums[end] >= target and nums[mid] <= target:
                    start = mid + 1
                else:
                    end = mid - 1
        return - 1

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
// If we know the start, we can assign left and right specific values to start.
class Solution {
    public int search(int[] nums, int target) {
        if(nums == null || nums.length == 0) return -1;
        int l = 0, r = nums.length-1, m = 0;
        // Find out the index of the smallest element.
        while(l < r) {
            m = (l + r) / 2;
            if(nums[m] > nums[r]) {
                l = m + 1;
            }else {
                r = m;
            }
        }

        // Since we now know the start, find out if the target is to left or right of start in the array.
        int s = l;
        l = 0; r = nums.length - 1;
        if(target >= nums[s] && target <= nums[r]) {
            l = s;
        }else {
            r = s;
        }
        // The regular search
        while(l <= r) {
            m = (l + r) / 2;
            if(nums[m] == target) return m;
            else if(nums[m] > target) r = m - 1;
            else l = m + 1;
        }

        return -1;
    }
}